Feature/recursive videos to poses

cleong110 opened the pull request:
Implemented a recursive option, solves #125 . Also added in an optional argument to specify the video extension to search for, though currently the only choice is .mp4. All behavior should be the same as before, as in if you don't use the -r argument it will just look in the current directory.

cleong110 commented:
Did a quick edit to the file to _SUPPORTED_VIDEO_FORMATS to see how easy it would be to switch extensions, and it successfully found and ran on .webm files as expected. But .webm support is a separate issue.  https://github.com/sign-language-processing/pose/issues/127

AmitMY commented on docs/specs/v0.1.md:
```diff
@@ -24,10 +24,16 @@
 \[`unsigned short` Blue]
 
+
+[Testing comment syntax from https://stackoverflow.com/a/32190021]: # 
```
why this test?

cleong110 replied:
Ah, leftover bit.

AmitMY commented on docs/specs/v0.1.md:
```diff
@@ -24,10 +24,16 @@
+\[`unsigned short` FPS]
+
+
+[THIS IS A PROBLEM]: # 
```
yes. this is fixed in v0.2

AmitMY commented on src/python/pose_format/bin/directory.py:
```diff
@@ -1,45 +1,42 @@
-            missing_pose_files.append(os.path.join(directory, mp4_file))
-
-    return sorted(missing_pose_files)
+    for vid_file in vid_files:
```
it is much faster to check all of the pose files
```py
glob_method = getattr (directory, 'rglob' if recursive else 'glob')
video_files = glob_method(f"*{video_suffix}")
pose_files = glob_method(f"*.pose")
```

AmitMY commented on src/python/pose_format/bin/directory.py:
```diff
@@ -1,45 +1,42 @@
-    parser.add_argument("--directory", type=str, required=True)
+    parser.add_argument("-d","--directory", type=Path, required=True, help="Directory to search for videos in")
+    parser.add_argument("-r", "--recursive", action="store_true", help="Whether to search for videos recursively")
+    parser.add_argument("--video_suffix", type=str, 
```
write `video-suffix` (dash, no underscore)

AmitMY commented on src/python/pose_format/bin/directory.py:
```diff
@@ -1,45 +1,42 @@
+    parser.add_argument("-r", "--recursive", action="store_true", help="Whether to search for videos recursively")
+    parser.add_argument("--video_suffix", type=str, 
+                        choices=_SUPPORTED_VIDEO_FORMATS,
+                        default=".mp4", help="Video extension to search for")
```
how about by default, we just go through all formats? something like `{'.mp4', '.mov', '.avi', '.mkv', '.flv', '.wmv', '.webm'}`

cleong110 replied:
Sure, that makes sense to me

cleong110 replied:
I was going for minimal behavior changes in this particular PR, but I do think just checking them all seems best

cleong110 commented:
Updated code to check against list of pose files as requested, and to iterate over all video suffixes. Realized this could result in foo.mp4, foo.webm both being posed to foo.pose, so added an option to keep original video suffix, and added a "this will cause name collisions do you want to keep going" check

cleong110 commented:
Also added a try/except checking for ValueError, so that if one of the video files is corrupted, the utility will keep going.

Also added a check for existing .pose files, in which case it will not re-run/overwrite.

AmitMY commented on docs/specs/v0.1.md:
```diff
@@ -24,10 +24,13 @@
+\[`unsigned short` FPS]
+
+
+\[`unsigned short` Number of frames] 
```
no need to modify the spec

cleong110 replied:
Should I put it back in?

AmitMY replied:
yeah i would revert the change to this file, no need to change it for this PR and there is just a comment removed

cleong110 replied:
Done!

AmitMY commented on src/python/pose_format/bin/directory.py:
```diff
@@ -1,45 +1,171 @@
 from tqdm import tqdm
 
+# Note: untested other than .mp4. Support for .webm may have issues: https://github.com/sign-language-processing/pose/pull/126
+_SUPPORTED_VIDEO_FORMATS = [".mp4", ".mov", ".avi", ".mkv", ".flv", ".wmv", ".webm"]
```
not a private variable, can be just uppercase, no initial underscore

AmitMY commented on src/python/pose_format/bin/directory.py:
```diff
@@ -1,45 +1,171 @@
+    glob_method = getattr(directory, "rglob" if recursive else "glob")
+    all_files = list(glob_method(f"*"))
+    video_files = [path for path in all_files if path.suffix in video_suffixes]
+    pose_files = [path for path in all_files if path.suffix == ".pose"]
```
pose files should be a `set` to make the check fast

AmitMY commented on src/python/pose_format/bin/directory.py:
```diff
@@ -1,45 +1,171 @@
 
+    for vid_path in video_files:
+        if (
+            not get_corresponding_pose_path(
```
the if statement is ugly. use a variable `corresponding_pose_path = get....` and then check

AmitMY commented on src/python/pose_format/bin/directory.py:
```diff
@@ -1,45 +1,171 @@
+    """
+    if keep_video_suffixes:
+        return video_path.with_name(f"{video_path.name}.pose")
+    else:
```
no need for `else`

AmitMY commented on src/python/pose_format/bin/directory.py:
```diff
@@ -1,45 +1,171 @@
+
+    print(f"Found {len(videos_with_missing_pose_files)} videos missing pose files.")
+
+    pose_files_that_will_be_created = set(
```
if you do `{... for ... in ...}` that is a set, no need to create a list and cast a set

cleong110 replied:
Huh! TIL

AmitMY commented on src/python/pose_format/bin/directory.py:
```diff
@@ -1,45 +1,171 @@
+        try:
+            pose_path = get_corresponding_pose_path(
+                video_path=vid_path, keep_video_suffixes=args.keep_video_suffixes
+            )
```
what's with all these line breaks? i feel like most of the lines would fit under 140 characters, but they are broken in most of the file

cleong110 replied:
I ran `black` on it, it must have been just over the threshold.

cleong110 replied:
Readjusted manually, then re-ran black with --line-length 140

AmitMY commented on src/python/pose_format/bin/directory.py:
```diff
@@ -1,45 +1,171 @@
+                print(f"Skipping {vid_path}, corresponding .pose file already created.")
+                continue
+            pose_video(vid_path, pose_path, args.format, additional_config)
+            pose_with_no_errors.append(vid_path)
```
if this is basically a counter, just count - `+=1` no need to create another array

cleong110 replied:
I had thought to use it at a later date for more detailed logging/debugging, e.g. write out the specific ones that didn't work. But I can just add the list then when/if it's needed.

AmitMY reviewed (changes requested):
much better. still minor changes

AmitMY reviewed (approved):
LGTM - just revert one change and i'll merge.
Thanks!

J22Melody commented:
It is great! 

I will add a multiprocessing option to the script to use multiple CPUs. Currently, I need to run pose estimation on 1000+ hours of BSL data, and do not want to wait for too long ;)

cleong110 commented:
https://github.com/cleong110/ud-vlab/blob/main/data_munging/split_folder.py
https://github.com/cleong110/ud-vlab/blob/main/data_munging/run_videos_to_poses_on_multiple_dirs.sh

J22Melody commented:
@cleong110 please share the repo with me!

cleong110 commented:
Oh silly me! The two files in question:
[run_videos_to_poses_on_multiple_dirs.zip](https://github.com/user-attachments/files/18027016/run_videos_to_poses_on_multiple_dirs.zip)
